catalog/rest_test.go

```diff
@@ -0,0 +1,818 @@
+			"defaults":  map[string]any{},
+			"overrides": map[string]any{},
+		})
+	})
+
+	r.srv = httptest.NewServer(r.mux)
+}
+
+func (r *RestCatalogSuite) TearDownTest() {
+	r.srv.Close()
+	r.srv = nil
+	r.mux = nil
+	r.configVals = nil
+}
+
+func (r *RestCatalogSuite) TestToken200() {
+	r.mux.HandleFunc("/v1/oauth/tokens", func(w http.ResponseWriter, req *http.Request) {
+		r.Equal(http.MethodPost, req.Method)
+
+		r.Equal(req.Header.Get("Content-Type"), "application/x-www-form-urlencoded")
+
+		r.Require().NoError(req.ParseForm())
+		values := req.PostForm
+		r.Equal(values.Get("grant_type"), "client_credentials")
+		r.Equal(values.Get("client_id"), "client")
+		r.Equal(values.Get("client_secret"), "secret")
+		r.Equal(values.Get("scope"), "catalog")
+
+		w.WriteHeader(http.StatusOK)
+
+		json.NewEncoder(w).Encode(map[string]any{
+			"access_token":      TestToken,
+			"token_type":        "Bearer",
+			"expires_in":        86400,
+			"issued_token_type": "urn:ietf:params:oauth:token-type:access_token",
+		})
+	})
+
+	cat, err := catalog.NewRestCatalog("rest", r.srv.URL,
+		catalog.WithWarehouseLocation("s3://some-bucket"),
+		catalog.WithCredential(TestCreds))
+	r.Require().NoError(err)
+
+	r.NotNil(cat)
+	r.Equal(r.configVals.Get("warehouse"), "s3://some-bucket")
+}
+
+func (r *RestCatalogSuite) TestToken400() {
+	r.mux.HandleFunc("/v1/oauth/tokens", func(w http.ResponseWriter, req *http.Request) {
+		r.Equal(http.MethodPost, req.Method)
+
+		r.Equal(req.Header.Get("Content-Type"), "application/x-www-form-urlencoded")
+
+		w.WriteHeader(http.StatusBadRequest)
+
+		json.NewEncoder(w).Encode(map[string]any{
+			"error":             "invalid_client",
+			"error_description": "credentials for key invalid_key do not match",
+		})
+	})
+
+	cat, err := catalog.NewRestCatalog("rest", r.srv.URL, catalog.WithCredential(TestCreds))
+	r.Nil(cat)
+
+	r.ErrorIs(err, catalog.ErrRESTError)
+	r.ErrorIs(err, catalog.ErrOAuthError)
+	r.ErrorContains(err, "invalid_client: credentials for key invalid_key do not match")
+}
+
+func (r *RestCatalogSuite) TestToken200AuthUrl() {
+	r.mux.HandleFunc("/auth-token-url", func(w http.ResponseWriter, req *http.Request) {
+		r.Equal(http.MethodPost, req.Method)
+
+		r.Equal(req.Header.Get("Content-Type"), "application/x-www-form-urlencoded")
+
+		r.Require().NoError(req.ParseForm())
+		values := req.PostForm
+		r.Equal(values.Get("grant_type"), "client_credentials")
+		r.Equal(values.Get("client_id"), "client")
+		r.Equal(values.Get("client_secret"), "secret")
+		r.Equal(values.Get("scope"), "catalog")
+
+		w.WriteHeader(http.StatusOK)
+
+		json.NewEncoder(w).Encode(map[string]any{
+			"access_token":      TestToken,
+			"token_type":        "Bearer",
+			"expires_in":        86400,
+			"issued_token_type": "urn:ietf:params:oauth:token-type:access_token",
+		})
+	})
+
+	authUri, err := url.Parse(r.srv.URL)
+	r.Require().NoError(err)
+	cat, err := catalog.NewRestCatalog("rest", r.srv.URL,
```

Comment:
same as above, this needs to verify the authorization header

Reply:
Since we test the authorization header in the `TestListTablesPrefixed200` I didn't think it necessary to add that same test in these. particularly given that we don't expose the client/session from the catalog object. I can add a test that isn't in the `_test` package so that I can verify the header though if you think we still need it.

Reply:
we're calling a different endpoint here, so I think we should still make sure that the authorization header is properly configured

Reply:
added test